fix scheduling metrics error for E2eSchedulingLatency

ddongchen opened the pull request:
If we get the scheduling metrics, which is "SchedulingAlgorithmLatency, E2eSchedulingLatency, BindingLatency". The E2eSchedulingLatency should be the sum of SchedulingAlgorithmLatency and BindingLatency, while we found E2eSchedulingLatency is almost the same as E2eSchedulingLatency for some optimization.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41780)
<!-- Reviewable:end -->

davidopp commented:
@ddongchen please follow the instructions about signing the CLA

bsalamat commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -124,7 +124,9 @@ func (s *Scheduler) scheduleOne() {
 
 	go func() {
-		defer metrics.E2eSchedulingLatency.Observe(metrics.SinceInMicroseconds(start))
+		defer func() {		
```
Could you please explain how this change addresses the issue?
I think you should pass `start` as an argument to the go routine at line 126 to solve the problem.

wojtek-t replied:
I don't understand this comment. I think the problem is not with start, but with that we are computing parameter of the function (so SinceInMicroseconds) here before doing binding.

bsalamat replied:
My bad. For some reason I thought the go routine and `start` were in a loop and thought that `start` changed before the go routine had the chance to run.

ddongchen replied:
The "E2eSchedulingLatency" should be calculated after the "BindingLatency" operation, by the "**defer metrics.E2eSchedulingLatency.Observe(metrics.SinceInMicroseconds(start))**". While in fact, the "E2eSchedulingLatency" is calculated before the "BindingLatency".  There may be some optimization for code compile

WIZARD-CXY replied:
because the defer function parameter is calculated first, it is just a rule, not code optimization

bsalamat commented:
/lgtm

ddongchen commented:
I signed it!

ddongchen commented:
I signed it!

ddongchen commented:
I signed it!

ddongchen commented:
@k8s-bot ok to test

wojtek-t commented:
```
I0329 12:02:33.060] Verifying hack/make-rules/../../hack/verify-gofmt.sh
I0329 12:02:43.364] !!! 'gofmt -s -w' needs to be run on the following files: 
I0329 12:02:43.364] ./plugin/pkg/scheduler/scheduler.go
I0329 12:02:43.366] [0;31mFAILED[0m   hack/make-rules/../../hack/verify-gofmt.sh	10s
```

@ddongchen - please fix the gofmt error and I will approve

ddongchen commented:
@k8s-bot ok to test

ddongchen commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot test this

wojtek-t commented:
/lgtm
/approve

wojtek-t commented:
Actually, @ddongchen - is this expected that you've send this PR to then 1.5 branch?
If you want it to be cherrypicked to 1.5 branch, the pattern is that:
- we first merge it to master branch
- only then do cherrypicks to other branches

ddongchen committed:
fix scheduling metrics error for E2eSchedulingLatency

If we get the scheduling metrics, which is "SchedulingAlgorithmLatency, E2eSchedulingLatency, BindingLatency". The E2eSchedulingLatency should be the sum of SchedulingAlgorithmLatency and BindingLatency, while we found E2eSchedulingLatency is almost the same as E2eSchedulingLatency for some optimization.

ddongchen commented:
@wojtek-t. I have changed to master! Sorry for this mistake, Please have a check again.

wojtek-t commented:
/lgtm

wojtek-t commented:
@k8s-bot test this

wojtek-t commented:
@k8s-bot kops test this

wojtek-t commented:
@k8s-bot non-cri e2e test this

WIZARD-CXY commented:
@davidopp I think this fix should be backported to v1.5 and v1.6 as well not just 1.7
